### PR TITLE
[PoC] Workflow task middleware feature

### PR DESCRIPTION
**Pull request by cgillum**
## What was changed
This PR adds a new feature called "workflow task middleware", which allows developers to wrap workflow task activations in middleware components.

The implementation is derived from the most basic implementation of ASP.NET middleware, and mimics a feature I created for the [Durable Task Framework](https://github.com/Azure/durabletask).

Note that this feature is distinct from workflow interceptors because the middleware runs _outside_ the workflow thread. This allows workflow code to run async. However, it's still subject to the deadlock detection logic in the WorkflowWorker (default of 2 seconds).

This is an advanced feature which most developers should not need. Rather, it's designed more for those who want to integrate the Temporal worker into other runtimes (such as the Azure Functions runtime, see below).

## Why?
This is for a PoC of having Azure Functions triggers and bindings for Temporal workflows.

For my specific case, I'm using this feature to create an Azure Functions extension that allows users to run Temporal workflows and activities serverless-ly and in a way that's natively understood by the Azure Functions platform. Specifically, each workflow and activity invocation (or "task", or even "activation", depending on which terminology you want to use) needs to be treated as a separate function invocation for the purposes of Azure Functions-specific timeout management, metering, billing, logging, etc.  

In summary, this feature allows me to wrap Temporal workflow activations in the Azure Functions execution pipeline. It's not needed for activities, only for workflows.

## Checklist

1. Closes (no issue opened)

2. How was this tested:
Manually, with both a simple console app and a real Azure Function app

3. Any docs updates needed?
Not unless we want to actually ship this.

### CODE REVIEW

**Comment by cretz** [comment]
I was originally wondering if we wanted to expose `WorkflowInstanceFactory`, expose `IWorkflowInstance`, and provide exposed wrappers around the otherwise not-exposed activation and activation completion types.

But I think this approach can work too, but would like to conform to the interceptor pattern. So obviously we make every call in `WorkflowInboundInterceptor` be in the workflow context. But we do create interceptors per workflow invocation so users can store state on the interceptors and therefore may not need a context do to that for them which is nice.

So, how about this...

Add this to `IWorkerInterceptor`:

```csharp
        WorkflowActivationInterceptor InterceptWorkflowActivations(WorkflowActivationInterceptor nextInterceptor) =>
            nextInterceptor;
```

Then these classes:

```csharp
public abstract class WorkflowActivationInterceptor
{
    // Basic interceptor stuff
    protected WorkflowActivationInterceptor(WorkflowActivationInterceptor next) => MaybeNext = next;
    private protected WorkflowActivationInterceptor() { }
    protected WorkflowActivationInterceptor Next => MaybeNext ?? throw new InvalidOperationException("No next interceptor");
    private protected WorkflowActivationInterceptor? MaybeNext { get; init; }

    // What you would want to override
    public virtual Task<WorkflowActivateOutput> ActivateAsync(WorkflowActivateInput input) => Next.ActivateAsync(input);
}

public record WorkflowActivateInput
{
    internal WorkflowActivateInput(
        WorkflowInfo workflowInfo,
        WorkflowDefinition workflowDefinition,
        WorkflowActivation workflowActivation)
    {
        WorkflowInfo = workflowInfo;
        WorkflowDefinition = workflowDefinition;
        WorkflowActivation = workflowActivation;
    }

    public WorkflowInfo WorkflowInfo { get; private init; }
    
    public WorkflowDefinition WorkflowDefinition { get; private init; }

    internal WorkflowActivation WorkflowActivation { get; init; }
}

public record WorkflowActivateOutput
{
    internal WorkflowActivateOutput(WorkflowActivationCompletion workflowActivationCompletion) =>
        WorkflowActivationCompletion = workflowActivationCompletion;

    internal WorkflowActivationCompletion WorkflowActivationCompletion { get; init; }
}
```

So, you could write an worker interceptor like this:

```csharp
public class MyWorkerInterceptor : IWorkerInterceptor
{
    private readonly string lookSomeExternalState;

    public MyWorkerInterceptor(string someState) => lookSomeExternalState = someState;

    // Invoked per workflow instance
    public WorkflowActivationInterceptor InterceptWorkflowActivations(WorkflowActivationInterceptor nextInterceptor) =>
        new MyWorkflowActivationInterceptor(lookSomeExternalState, nextInterceptor);

    private class MyWorkflowActivationInterceptor : WorkflowActivationInterceptor
    {
        // The state survives the life of the workflow instance. But of course
        // not the workflow run which may resume on some completely separate
        // process.
        private readonly string lookSomeConstructedState;
        private readonly Dictionary<string, object?> lookMutableState = new();

        internal MyWorkflowActivationInterceptor(string someState, WorkflowInboundInterceptor next)
            : base(next) => lookSomeConstructedState = someState;

        public override async Task<WorkflowActivateOutput> ActivateAsync(WorkflowActivateInput input)
        {
            // You can trust this is a thread-safe access because we
            // guarantee serial activation invocations
            lookMutableState["foo"] = "bar";
            try
            {
                return await base.ActivateAsync(input);
            }
            finally
            {
                Console.WriteLine("I'm running after every execution, success or fail")
            }
        }
    }
}
```

If you agree with this approach (which is basically what you have, just some different names and state management approaches), I will fire up a PR to the repo with it.
